### PR TITLE
feat(*): Webpack 4 Support for Universal CSS Chunks

### LINKED ISSUE #60: webpack 4 support
Hey there @faceyspacey 
Hope you are doing well.

I was messing around with the next version of webpack i.e `4.0.0-beta.0` and ran into an issue when trying to compile with it.

```
Error: Chunk.entrypoints: Use Chunks.groupsIterable and filter by instanceof Entrypoint instead
    at Chunk.get (/home/guillaume/Projects/guillaumecisco/node_modules/webpack/lib/Chunk.js:463:9)
    at /home/guillaume/Projects/guillaumecisco/node_modules/extract-css-chunks-webpack-plugin/index.js:243:40
    at Array.forEach (<anonymous>)
    at ExtractTextPlugin.<anonymous> (/home/guillaume/Projects/guillaumecisco/node_modules/extract-css-chunks-webpack-plugin/index.js:238:11)
    at AsyncSeriesHook.eval [as callAsync] (eval at create (/home/guillaume/Projects/guillaumecisco/node_modules/tapable/lib/HookCodeFactory.js:24:12), <anonymous>:7:1)
    at AsyncSeriesHook.lazyCompileHook [as _callAsync] (/home/guillaume/Projects/guillaumecisco/node_modules/tapable/lib/Hook.js:35:21)
    at Compilation.seal (/home/guillaume/Projects/guillaumecisco/node_modules/webpack/lib/Compilation.js:789:27)
    at hooks.make.callAsync.err (/home/guillaume/Projects/guillaumecisco/node_modules/webpack/lib/Compiler.js:439:17)
    at _err0 (eval at create (/home/guillaume/Projects/guillaumecisco/node_modules/tapable/lib/HookCodeFactory.js:24:12), <anonymous>:11:1)
    at _addModuleChain (/home/guillaume/Projects/guillaumecisco/node_modules/webpack/lib/Compilation.js:672:11)
    at processModuleDependencies.err (/home/guillaume/Projects/guillaumecisco/node_modules/webpack/lib/Compilation.js:614:8)
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickCallback (internal/process/next_tick.js:180:9)
```

Looks like the exact same problem as https://github.com/webpack-contrib/extract-text-webpack-plugin/issues/701

Let me know if I can help further.
:call_me_hand:

### PR DESCRIPTION
**Pull request by zackljackson**
Webpack 4 support for extract-css-chunks universal-component react-universal. Complete with full
chunking, code splitting, real and true hot module reloading. Compatible with faceyspacey universal
code splitting and server-side rendering.

Pretty much its great chunking like before, with really slick HMR. Putting it all together. The only plugin that does both css code splitting and HMR 👍 
From Readme:

## Perks
* **HMR:** It also has first-class support for **Hot Module Replacement** across ALL those css files/chunks!!!
* cacheable stylesheets 
* smallest total bytes sent compared to "render-path" css-in-js solutions that include your CSS definitions in JS
* Faster than the V2!
* Async loading
* No duplicate compilation (performance)
* Easier to use
* Specific to CSS
* SSR Friendly development build, focused on frontend DX
* Works seamlessly with the Universal family
* Works fantastically as a standalone style loader (You can use it for any webpack project! with no extra dependencies!)

BREAKING CHANGE: EVERYTHING! This its a complete rewrite. Top to bottom!! This is a major change to the API and Loader. It will intergrate with universal without a issue and  IT DOES NOT RELY ON
UNIVERSAL COMPONENTS.

fix #60, re #36 #47 #51

### CODE REVIEW

**Comment by zackljackson** [comment]
I need to test this with the rest of universal and start pushing those for PR if needed

File: src/index.js [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+    }
+  }
+
+  apply(compiler) {
+    if (process.env.NODE_ENV === 'development') {
+      try {
+        const isHOT = isHMR(compiler);
+
+        if (isHOT && compiler.options.module && compiler.options.module.rules) {
+          const updatedRules = compiler.options.module.rules.reduce((rules, rule) => {
```
Comment (zackljackson) [inline src/index.js]: @faceyspacey any ideas man. Its stable but looks off to me

Comment (faceyspacey) [reply src/index.js]: what looks off? its automatically adding the hotloader if its using the css loader.

Comment (zackkirsh) [reply src/index.js]: Essentially yes, is this the most optimal method of accessing the compiler's loaders? Other than that, im fine with it. Just checking if you know about something like `prependLoader`. If not then all good

**Comment by faceyspacey** [comment]
oh, and if it's not clear, FUCKIN ROUND OF APPLAUSE! Well done @zackljackson

File: src/index.js (hunk 1 above)
Comment (faceyspacey) [reply src/index.js]: na, i've seen this pattern of rewriting the rules at least once or twice before.

**Comment by zackljackson** [comment]
So I was thinking about opening a PR back to mini-css extract - but theres a lot of people who already use and depend on our repo. It would be good to give it a webpack 4 upgrade, so the users can just update. 

I'd also like to keep a solution under our domain in the event that I need more than mini-css can offer. 
Honestly I can't think of a good reason right now, but it would be nice to keep a heavily depended upon plugin updated. 

I do want to open this to mini-css, but id also like to release it here first

**Comment by faceyspacey** [comment]
ok. proceed. all this stuff is intended for cross-polination anyway. this can be the breeding ground. ..but also, i think mini extract itself is a stopgap and webpack is supposed to have full built-in support for css or something. in other words: webpack's css solution is likely to change, so i think it's a good thing that we stay steady continuing to fulfill our traditional role.

**Comment by zackljackson** [comment]
Agreed. Ill open a PR to them soon - but as you said mini-css is totally a stopgap. It'll be good to maintain control of our architecture till a stable replacement is made ready in webpack

**Pull request merged**